src/colmap/estimators/essential_matrix.cc

```diff
@@ -176,38 +174,39 @@ void EssentialMatrixEightPointEstimator::Estimate(
   CenterAndNormalizeImagePoints(points2, &normed_points2, &normed_from_orig2);
 
   // Setup homogeneous linear equation as x2' * F * x1 = 0.
-  Eigen::Matrix<double, Eigen::Dynamic, 9> cmatrix(points1.size(), 9);
+  Eigen::Matrix<double, Eigen::Dynamic, 9> A(points1.size(), 9);
   for (size_t i = 0; i < points1.size(); ++i) {
-    cmatrix.block<1, 3>(i, 0) = normed_points1[i].homogeneous();
-    cmatrix.block<1, 3>(i, 0) *= normed_points2[i].x();
-    cmatrix.block<1, 3>(i, 3) = normed_points1[i].homogeneous();
-    cmatrix.block<1, 3>(i, 3) *= normed_points2[i].y();
-    cmatrix.block<1, 3>(i, 6) = normed_points1[i].homogeneous();
+    A.row(i) << normed_points2[i].x() *
+                    normed_points1[i].transpose().homogeneous(),
+        normed_points2[i].y() * normed_points1[i].transpose().homogeneous(),
+        normed_points1[i].transpose().homogeneous();
   }
 
   // Solve for the nullspace of the constraint matrix.
-  Eigen::JacobiSVD<Eigen::Matrix<double, Eigen::Dynamic, 9>> cmatrix_svd(
-      cmatrix, Eigen::ComputeFullV);
-  const Eigen::VectorXd ematrix_nullspace = cmatrix_svd.matrixV().col(8);
-  const Eigen::Map<const Eigen::Matrix3d> ematrix_t(ematrix_nullspace.data());
-
-  // De-normalize to image points.
-  const Eigen::Matrix3d E_raw =
-      normed_from_orig2.transpose() * ematrix_t.transpose() * normed_from_orig1;
+  Eigen::Matrix3d Q;
+  if (points1.size() == 8) {
+    Eigen::Matrix<double, 9, 9> QQ =
+        A.transpose().householderQr().householderQ();
+    Q = Eigen::Map<const Eigen::Matrix<double, 3, 3, Eigen::RowMajor>>(
+        QQ.col(8).data());
+  } else {
+    Eigen::JacobiSVD<Eigen::Matrix<double, Eigen::Dynamic, 9>> svd(
+        A, Eigen::ComputeFullV);
+    Q = Eigen::Map<const Eigen::Matrix<double, 3, 3, Eigen::RowMajor>>(
+        svd.matrixV().col(8).data());
+  }
 
-  // Enforcing the internal constraint that two singular values must be equal
+  // Enforcing the internal constraint that two singular values must non-zero
   // and one must be zero.
-  Eigen::JacobiSVD<Eigen::Matrix3d> E_raw_svd(
-      E_raw, Eigen::ComputeFullU | Eigen::ComputeFullV);
-  Eigen::Vector3d singular_values = E_raw_svd.singularValues();
-  singular_values(0) = (singular_values(0) + singular_values(1)) / 2.0;
-  singular_values(1) = singular_values(0);
+  Eigen::JacobiSVD<Eigen::Matrix3d> svd(
+      Q, Eigen::ComputeFullU | Eigen::ComputeFullV);
+  Eigen::Vector3d singular_values = svd.singularValues();
   singular_values(2) = 0.0;
-  const Eigen::Matrix3d E = E_raw_svd.matrixU() * singular_values.asDiagonal() *
-                            E_raw_svd.matrixV().transpose();
+  const Eigen::Matrix3d E =
+      svd.matrixU() * singular_values.asDiagonal() * svd.matrixV().transpose();
 
   models->resize(1);
-  (*models)[0] = E;
+  (*models)[0] = normed_from_orig2.transpose() * E * normed_from_orig1;
```

Comment:
Not sure how this ended up wrongly implemented, but the previous implementation was just wrong.

Reply:
Should we change the matches generation for the unit tests to 3D and compute essential matrix from the relative pose. In this way we can avoid the previous issue of wrong implementation. Now we can only compare with the reference : )

Reply:
I probably misunderstand your proposal, but there is a new "EssentialMatrixEightPointEstimatorTests" test that does that.

Reply:
Thanks! I did mean exactly this